src/payload.rs

```diff
@@ -0,0 +1,269 @@
+//! You can find more information about the protocol v1 in the [Astarte MQTT v1 Protocol](https://docs.astarte-platform.org/astarte/latest/080-mqtt-v1-protocol.html).
+
+use std::collections::HashMap;
+
+use bson::Bson;
+use chrono::{DateTime, Utc};
+use log::trace;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    types::{AstarteType, TypeError},
+    Aggregation,
+};
+
+/// Errors that can occur handling the payload.
+#[non_exhaustive]
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum PayloadError {
+    /// Couldn't serialize the payload to bson.
+    #[error("couldn't serialize the payload")]
+    Serialize(#[from] bson::ser::Error),
+    /// Couldn't deserialize the payload to bson.
+    #[error("couldn't deserialize the payload")]
+    Deserialize(#[from] bson::de::Error),
+
+    /// Couldn't convert the value to [`AstarteType`]
+    #[error("couldn't convert the value to AstarteType")]
+    AstarteType(#[from] TypeError),
+}
+
+/// The payload of an MQTT message.
+///
+/// It is serialized as a BSON object when sent over the wire.
+///
+/// The payload BSON specification can be found here: [BSON](https://docs.astarte-platform.org/astarte/latest/080-mqtt-v1-protocol.html#bson)
+#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
+pub(crate) struct Payload<T> {
+    #[serde(rename = "v")]
+    pub(crate) value: T,
+    #[serde(rename = "t", default, skip_serializing_if = "Option::is_none")]
+    pub(crate) timestamp: Option<DateTime<Utc>>,
+}
+
+impl<T> Payload<T>
+where
+    T: serde::Serialize,
+{
+    pub(crate) fn to_vec(&self) -> Result<Vec<u8>, PayloadError> {
+        let res = bson::to_vec(self)?;
+
+        Ok(res)
+    }
+
+    pub(crate) fn from_slice<'a>(buf: &'a [u8]) -> Result<Payload<T>, PayloadError>
+    where
+        T: serde::de::Deserialize<'a>,
+    {
+        let res = bson::from_slice(buf)?;
+
+        Ok(res)
+    }
+}
+
+pub(crate) fn serialize_individual(
+    data: &AstarteType,
+    timestamp: Option<DateTime<Utc>>,
+) -> Result<Vec<u8>, PayloadError> {
+    let payload = Payload {
+        value: data,
+        timestamp,
+    };
+
+    payload.to_vec()
+}
+
+pub(crate) fn deserialize(bdata: &[u8]) -> Result<Aggregation, PayloadError> {
+    if bdata.is_empty() {
+        return Ok(Aggregation::Individual(AstarteType::Unset));
+    }
+
+    let payload = Payload::<Bson>::from_slice(bdata)?;
+
+    trace!("{:?}", payload);
+
+    match payload.value {
+        Bson::Document(doc) => {
+            let hmap = doc
+                .into_iter()
+                .map(|(name, value)| {
+                    let v = AstarteType::try_from(value)?;
+
+                    Ok((name, v))
+                })
+                .collect::<Result<HashMap<String, AstarteType>, PayloadError>>()?;
+
+            Ok(Aggregation::Object(hmap))
+        }
+        value => {
+            let individual = AstarteType::try_from(value)?;
+
+            Ok(Aggregation::Individual(individual))
+        }
+    }
+}
```

Comment:
Move this function below `fn serialize_object`.

Reply:
Done